StackInfo issues on Alpine (musl-based)

continuing from https://github.com/SerenityOS/ladybird/issues/153#issue-1509918847:

> it looks like the stack size returned by `StackInfo#size_free()` is invalid, causing constant `[InternalError] Call stack size limit exceeded` on even the smallest scripts. i tried to debug this, but sadly didn't get very far; also haven't tested if the same issue happens on gentoo or void with musl

> To confirm, this is when running ladybird in a musl-libc based desktop environment?

@ADKaster Correct, I've tried running it on multiple machines and all of them had the same issue.

I tried running the tests myself on 379e4a2432e3db557c4ffb47a57a08df40f5662f (HEAD of master at the moment of posting) and they fail with the following:
```
serenity/Userland/Utilities/ntpquery.cpp:214:13: error: comparison of integers of different signs: 'unsigned long' and 'long' [-Werror,-Wsign-compare]
    VERIFY(!CMSG_NXTHDR(&msg, cmsg));
    ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/sys/socket.h:358:44: note: expanded from macro 'CMSG_NXTHDR'
        __CMSG_LEN(cmsg) + sizeof(struct cmsghdr) >= __MHDR_END(mhdr) - (unsigned char *)(cmsg) \
                                                  ^
/usr/include/assert.h:8:28: note: expanded from macro 'assert'
#define assert(x) ((void)((x) || (__assert_fail(#x, __FILE__, __LINE__, __func__),0)))
                           ^
1 error generated.
```

--- a/AK/StackInfo.cpp
+++ b/AK/StackInfo.cpp
@@ -14,6 +14,7 @@
 #    include <serenity.h>
 #elif defined(AK_OS_LINUX) || defined(AK_LIBC_GLIBC) || defined(AK_OS_MACOS) || defined(AK_OS_IOS) || defined(AK_OS_NETBSD) || defined(AK_OS_SOLARIS) || defined(AK_OS_HAIKU)
 #    include <pthread.h>
+#    include <sys/resource.h>
 #elif defined(AK_OS_FREEBSD) || defined(AK_OS_OPENBSD)
 #    include <pthread.h>
 #    include <pthread_np.h>
@@ -95,6 +96,21 @@ StackInfo::StackInfo()
 #endif
 
     m_top = m_base + m_size;
+
+#if defined(AK_OS_LINUX) && !defined(AK_OS_ANDROID) && !defined(AK_LIBC_GLIBC)
+    // Note: musl libc always gives the initial size of the main thread's stack
+    if (getpid() == static_cast<pid_t>(gettid())) {
+        rlimit limit;
+        getrlimit(RLIMIT_STACK, &limit);
+        rlim_t size = limit.rlim_cur;
+        if (size == RLIM_INFINITY)
+            size = 8 * 0x10000;
+        // account for a guard page
+        size -= static_cast<rlim_t>(sysconf(_SC_PAGESIZE));
+        m_size = static_cast<size_t>(size);
+        m_base = m_top - m_size;
+    }
+#endif
 }
 
 }

